### ISSUE TITLE
Compiling with Fable in watch seems to output empty file before writting to them

### ISSUE THREAD

**Issue by MangelMaxime** [fable-compiler/Fable#2351]
When using Fable in watch mode `dotnet fable Tests.fsproj --watch` it seems that at some point the files are delete and then recreated even if there was no modification made.

This leads so strange situation when using a file watch like mocha.

![image](https://user-images.githubusercontent.com/4760796/104961478-b2a58f00-59d6-11eb-9b63-a6240e1e9ac4.png)

![image](https://user-images.githubusercontent.com/4760796/104961496-bafdca00-59d6-11eb-94b4-4d44f0a7649f.png)

For both screenshot, I just added a space to file before saving it (my VSCode trim whitespace) so the file didn't change. However, when looking at the file information I see that the creation date has been modified which means it is no more the "same file" I think.

----

In another project which use webpack for client side bundling. I saw some error about a missing export in some of the modules. I suspect this is caused by the same problem we see here. I think the watch see a file change decide to bundle but is too quick and sometime access an empty file or something like that.

@alfonsogarciacaro Does this kind of behavior as I described ring something to you?

**Comment by MangelMaxime** [comment]
Reproduction repo:

1. Clone the repo at this commit/branch https://github.com/MangelMaxime/Fable.Express/tree/2a9e13992c3a0f1bf33cd95bed251b94d81fdde2
2. `npm i`
3. `npm run watch-test`
4. Make some change on `tests/Qs/Qs.Tests.fs` and save the file
5. Observe that sometime mocha execute once sometimes several times with unexpected result (0 passing)

**Comment by alfonsogarciacaro** [comment]
Ah, now that you mention this, I think I've seen something similar with Mocha in watch mode. Please note that a "watch dependency tree" is being created and modification in some files trigger recompilation in others because even if the F# code hasn't changed, the JS code may need to (e.g. because the overload hash or the union tag has changed) in order to avoid issues like #2342

**Comment by MangelMaxime** [comment]
> Please note that a "watch dependency tree" is being created and modification in some files trigger recompilation in others because even if the F# code hasn't changed, the JS code may need to (e.g. because the overload hash or the union tag has changed) in order to avoid issues like #2342

Ok, I think this is fine especially if this makes Fable works reliably in the case you mention :)

I just think Fable should overwrite existing files and not delete / create them. Please note I didn't checked Fable code so perhaps it not deleted/created them but when looking at the file information this the suspicion I have.

Thinking of it, perhaps making Fable overwrite the file will only mitigates the issue because it's possible that Mocha/Bundler execute before while Fable is still writing some files.

Pseudo timeline:

- Fable write File A
- Mocha see changes in File A and re-execute
- Fable write File B
- Mocha is done with the execution
- Fable write File C / Fable done

In general, I use `nodemon --delay 250ms` to aggregate the changes in a single "watch step". I will see if I can make `nodemon` execute mocha for me because mocha doesn't seems to have this kind of feature.

----------------

*I am telling this from memory so it can be wrong/incomplete*

I think I had similar issues when using snowpack and rollup but not when using webpack.

After a quick search Webpack has an option [aggregateTimeout](https://webpack.js.org/configuration/watch/#watchoptionsaggregatetimeout) for this kind of situation which seems 200ms by default. That's probably why I didn't saw this problem when using webpack for a client app.

**Comment by MangelMaxime** [comment]
I tested using `npx nodemon --watch tests --exec 'mocha -r esm tests/Main.fs.js'` and it works.

Note that it doesn't introduce a delay and don't have the same problem as Mocha does. I suspect nodemon to have a better watch mode than Mocha.

It seems that the "issue" will depends on the bundler/runner used at the end of the pipeline. I don't know if there is something we can improve on Fable side or if that's just how life is.

**Comment by alfonsogarciacaro** [comment]
I'm just using a .NET IO Stream to write the files, not sure how it handles overwriting: https://github.com/fable-compiler/Fable/blob/10cb7d4eba3c8c8b22ac317d23ebdfb28a1a0ec5/src/Fable.Cli/Main.fs#L141-L144

I'm using the mocha watcher because the extra-run issue is not bothering me 😅 But sorry, forgot to tell you Fable includes a `--runWatch` option which is like `--run` but happens after every watch compilation (killing the previous process if there's one). I used it to replace nodemon when running a Express server compiled with Fable, probably you can do something similar with mocha.

```
dotnet fable watch src/Server -o build/server --runWatch node build/server --env=development
```

Just FYI, in the `build/server` folder I generate an `index.js` file to load ESM modules like this:

```js
require = require('esm')(module);
module.exports = require('./Main.js');
```

**Comment by MangelMaxime** [comment]
> I'm just using a .NET IO Stream to write the files, not sure how it handles overwriting: 

Thank you for the pointer, I don't know neither. I tried to search online but didn't found anything promising on the subject.

> I'm using the mocha watcher because the extra-run issue is not bothering me sweat_smile 

When running my tests I prefer to have a deterministic behavior just in case :) 

Ah yes I forgot about the `--runScript` it can be a solution for one of my project indeed.

I will close this issue as we listed several alternative solution to workaround the "problem". And also because we don't really have an idea of how to improve it as the code is "just" writing to the output file using .NET API.

Thank you for the help :)
